### PR TITLE
Feat: Add CVE support to Snyk datasource

### LINKED ISSUE #1163: Add CVE support in SnykDataSource
Implement `datasource_advisory_from_cve` in SnykDataSource

For example, see `GithubDataSource.datasource_advisory_from_cve`:
https://github.com/nexB/vulnerablecode/blob/0a8f7a863de3062deabcb3cff347f64bee1e776e/vulntotal/datasources/github.py#L54-L73

### PR DESCRIPTION
**Pull request by shravankshenoy**
Fixes #1163 

### Changes Made
Added function `datasource_advisory_from_cve` to Snyk data source and other helper functions. 

### Working
An overview of how the function works is given below
1. `generate_payload_from_cve` function returns url for a given cve
2. Html page from the given cve url is fetched using `fetch` function 
3. `parse_cve_advisory_html` function extracts all the vulnerabilities and their snyk_ids and package ids from html page
4. All the affected versions for the affected package id is obtained using `extract_html_json_advisories` function
5. Snyk url and html page are got using `generate_advisory_payload` and `fetch` functions
6. All the data is passed into `parse_html_advisory function` along with advisory_html page created in step 5, and that will return the final VendorData object with affected and fixed versions

`generate_purl` is a helper function which generates a purl from package url

### Results
Tested the app locally by printing the results yielded for CVE-2024-23641

image

This matches the result at https://security.snyk.io/vuln?search=CVE-2024-23641

### CODE REVIEW

**Comment by shravankshenoy** [comment]
Oops looks like one test related related to formatting is failing - Black style check failed. Will try to fix that

**Comment by ziadhany** [comment]
> Oops looks like one test related to formatting is failing - Black style check failed. Will try to fix that

@shravankshenoy you can just run `make valid` and check the test again with `make test`

**Comment by shravankshenoy** [comment]
Thanks @ziadhany . All the tests are passing now.

File: vulntotal/tests/test_snyk.py [hunk 1]
```diff
@@ -52,6 +52,37 @@ def test_generate_package_advisory_url(self):
         ]
         util_tests.check_results_against_expected(results, expected)
 
+    def test_generate_purl(self):
```
Comment (keshav-space) [inline vulntotal/tests/test_snyk.py:55]: Add a few tests for `golang` too.

**Review by keshav-space — CHANGES_REQUESTED** [review]
Thanks, @shravankshenoy. This is a good start.
Please take a look at the suggestions below.

File: vulntotal/datasources/snyk.py [hunk 2]
```diff
@@ -132,6 +166,53 @@ def generate_package_advisory_url(purl):
...
+
+    elif pkg_type == "linux":
+        pkg_name = package_url_split[-1]
+        qualifiers["distro"] = package_url_split[1]
+
+    else:
+        pkg_name = package_url_split[-1]
+
+    return PackageURL(
+        type=pkg_type, name=pkg_name, namespace=namespace, version=version, qualifiers=qualifiers
```
[the remark this answers is not in the document]
Comment (shravankshenoy) [reply vulntotal/datasources/snyk.py]: Done

File: vulntotal/datasources/snyk.py [hunk 3]
```diff
@@ -132,6 +166,53 @@ def generate_package_advisory_url(purl):
...
+            version = package_url_split[-1]
+        else:
+            pkg_name = package_url_split[-1]
+            namespace = package_url_split[-2]
+
+    elif pkg_type == "linux":
+        pkg_name = package_url_split[-1]
+        qualifiers["distro"] = package_url_split[1]
+
+    else:
```
[the remark this answers is not in the document]
Comment (shravankshenoy) [reply vulntotal/datasources/snyk.py]: Got it. I have added an additional if condition to handle scoped npm packages.

File: vulntotal/tests/test_snyk.py (hunk 1 above)
Comment (shravankshenoy) [reply vulntotal/tests/test_snyk.py:55]: Done. Have added additional tests for golang and scoped npm packages.

**Comment by shravankshenoy** [comment]
Thanks a lot for reviewing the code. It not only helped me understand purl-spec much better, but also helped while wrtitng code for some of the other PRs. As a positive side effect of this change, when I ran the code for CVE-2024-23641, it now shows the namespaces, unlike before where namespaces were null.

image

 Let me know if any further changes required, would be glad to the the needful :)

File: vulntotal/datasources/snyk.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (shravankshenoy) [reply vulntotal/datasources/snyk.py]: Thanks for sharing this. I have tried my best to make sure the purl-spec is followed properly now, but let me know if I missed something.

File: vulntotal/tests/test_snyk.py [hunk 4]
```diff
@@ -52,6 +52,52 @@ def test_generate_package_advisory_url(self):
...
+            "https://security.snyk.io/package/composer/bolt%2Fcore",
+            "https://security.snyk.io/package/linux/debain:11/trafficserver",
+            "https://security.snyk.io/package/linux/almalinux:8/rpm-plugin-fapolicyd",
+            "https://security.snyk.io/package/nuget/moment.js",
+            "https://security.snyk.io/package/cocoapods/ffmpeg",
+            "https://security.snyk.io/package/hex/coherence",
+            "https://security.snyk.io/package/rubygems/log4j-jars",
+            "https://security.snyk.io/package/golang/github.com%2Fgrafana%2Fgrafana%2Fpkg%2Fservices%2Fsqlstore%2Fmigrator",
+            "https://security.snyk.io/package/golang/github.com%2Fanswerdev%2Fanswer%2Finternal%2Frepo%2Factivity",
+            "https://security.snyk.io/package/golang/github.com%2Fargoproj%2Fargo-cd%2Fv2%2Futil%2Fhelm",
```
Comment (keshav-space) [inline vulntotal/tests/test_snyk.py:72]: For better test coverage, add few non-GitHub Go packages in the test too.

**Review by keshav-space — CHANGES_REQUESTED** [review]
@shravankshenoy see few suggestions below.

File: vulntotal/datasources/snyk.py [hunk 5]
```diff
@@ -132,6 +166,59 @@ def generate_package_advisory_url(purl):
...
+        pkg_name = package_url_split[-1]
+
+    elif pkg_type == "linux":
+        pkg_name = package_url_split[-1]
+        qualifiers["distro"] = package_url_split[1]
+
+    elif pkg_type in ("cocoapods", "hex", "nuget", "pip", "rubygems", "unmanaged"):
+        pkg_name = package_url_split[-1]
+
+    return PackageURL(type=pkg_type, name=pkg_name, namespace=namespace)
```
[the remark this answers is not in the document]
Comment (shravankshenoy) [reply vulntotal/datasources/snyk.py]: Done. Added an if condition which logs an error if type or name is missing. Also `datasource_advisory_from_cve` function has been modified such that if the `generate_purl` function returns None, the advisory is skipped

File: vulntotal/datasources/snyk.py [hunk 6]
```diff
@@ -132,6 +166,59 @@ def generate_package_advisory_url(purl):
...
+    elif pkg_type == "composer":
+        pkg_name = package_url_split[-1]
+        namespace = package_url_split[-2]
+
+    elif pkg_type == "golang":
+        if package_url_split[1] == "github.com":
+            ns_start = package_advisory_url.find("github.com")
+            ns_end = package_advisory_url.rfind("/")
+            namespace = package_advisory_url[ns_start:ns_end]
+            pkg_name = package_url_split[-1]
```
[the remark this answers is not in the document]
Comment (shravankshenoy) [reply vulntotal/datasources/snyk.py]: Done

File: vulntotal/tests/test_snyk.py (hunk 4 above)
Comment (shravankshenoy) [reply vulntotal/tests/test_snyk.py]: Done, added a few non Github Go packages

**Comment by shravankshenoy** [comment]
Thanks for reviewing the code. Made all the requested changes. 

While working on these changes, I also realized that I had somehow missed on following the purl spec properly. So I have used the dictionary which we get by calling `SnykDataSource.supported_ecosystem()`, inverted the keys and values, and use the new dictionary to get the purl-spec compliant package type. Hope that works.

**Review by keshav-space — APPROVED** [review]
Thanks++, LGTM!

**Pull request merged**